drivers/flash/flash_sam.c

```diff
@@ -142,10 +160,14 @@ static int flash_sam_write_page(const struct device *dev, off_t offset,
 	LOG_DBG("offset = 0x%lx, len = %zu", (long)offset, len);
 
 	/* We need to copy the data using 32-bit accesses */
-	for (; len > 0; len -= sizeof(*src)) {
-		*dst++ = *src++;
+	for (int i = 0; i < len / sizeof(*src); i++) {
+		dst[i] = src[i];
+		/* Assure data are written to the latch buffer consecutively */
+		__DSB();
 	}
-	__DSB();
+
+	/* Assure cache coherency before flash write operation */
+	dcache_clean(dst, len);
```

Comment:
Is this really necessary?

From ["How to manage Cortex-M7 Cache Coherence on the Atmel SAM S70 / E70"]():
![image](https://user-images.githubusercontent.com/11345418/128851521-9d133a8d-5521-4bd5-bcb0-d7cece0b0444.png)

According to the application note above, the flash region is write-through by default (i.e. even when you do not explicitly configure the caching policy through MPU), so DSB should be enough.

Reply:
Taking into account Write-Through, no Write-Allocate most likely not. I'll remove it.